### PR TITLE
feat!: add key pairs and accounts settings to wallet settings

### LINKED ISSUE #19936: [Epic] - Feature - Mobile Wallet Settings - Key Pairs and Accounts
All features related to issue for epic - [E:Mobile Wallet Settings - Key Pairs and Accounts](https://github.com/status-im/status-mobile/labels/E%3AMobile%20Wallet%20Settings%20-%20Key%20Pairs%20and%20Accounts)

### User Story
- As a user, I want to be able to manage my key pairs and accounts from within the wallet settings screen. 
- I want to be able to rename my key pairs
- I want to be able to remove my key pairs
- I want to be able to share my key pairs via a QR code
- I want to be able to view my missing key pairs
- I want to be able to import a missing key pair by its recovery phrase
- I want to be able to import a missing key pair by its private key
- I want to be able to import a missing key pair by a QR code
- I want to be able to import all missing key pairs by a QR code

**Related Designs**:
* Overview of the de-scoped Key Pairs and Accounts flows: [the Figma designs](https://www.figma.com/file/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?type=design&node-id=259-24408&mode=design&t=M9zZRtqh1MBl3m5j-4)
* Overview of the de-scoped Missing Key Pairs flows: [the Figma designs](https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=232-32244&)

**Features**:
These features should be hidden behind a feature flag.

Mobile Wallet Settings: implement the UI for displaying key-pairs and accounts:
- [x] https://github.com/status-im/status-mobile/issues/19916

Mobile Wallet Settings: implement the UI for renaming a key-pair:
- [x] https://github.com/status-im/status-mobile/issues/19917

Mobile Wallet Settings: implement the UI for removing a key-pair:
- [x] https://github.com/status-im/status-mobile/issues/19921

Mobile Wallet Settings: implement the UI for sharing a key-pair:
- [x] https://github.com/status-im/status-mobile/issues/19946

Mobile Wallet Settings: implement the UI for displaying missing key-pairs:
- [x] https://github.com/status-im/status-mobile/issues/20033

Mobile Wallet Settings: implement the UI for importing a missing key-pair with a recovery phrase:
- [x] https://github.com/status-im/status-mobile/issues/20037

Mobile Wallet Settings: implement the UI for importing a missing key-pair with a private key:
- [x] https://github.com/status-im/status-mobile/issues/20374

Mobile Wallet Settings: implement the UI for importing a missing key-pair with a QR-code:
- [x] https://github.com/status-im/status-mobile/issues/20165

Mobile Wallet Settings: implement the UI for importing all missing key-pairs with a QR-code:
- [x] https://github.com/status-im/status-mobile/issues/20035

**Closing this issue**
Once all of these issues are ready the corresponding feature Flag Can be removed. This pr to remove the feature flag can then be used for QA and Design to ensure everything is feature complete and this issue can be closed.

### PR DESCRIPTION
**Pull request by seanstrom**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

fixes #19936 

### Summary

* This PR attempts to reveal the feature-flagged "Key pairs and accounts" settings inside wallet settings

### Review notes

* A summary of the related features that need to tested and reviewed can be found in this epic: https://github.com/status-im/status-mobile/issues/19936 

### Testing notes

* Keep in mind that the testing environment will need to create a missing key pairs, which requires syncing two devices and creating a new key-pair on one device (Device A) and viewing that key-pair as missing on the other device (Device B).

#### Platforms

- Android
- iOS

##### Functional

- Settings
- Wallet Settings
- Key pairs and accounts settings
- Missing key pairs

### Steps to test

* Steps to test this feature can be found in the related issues that described in this epic: https://github.com/status-im/status-mobile/issues/19936 

status: ready

### CODE REVIEW

File: src/status_im/contexts/settings/wallet/wallet_options/view.cljs
Comment (smohamedjavid) [inline src/status_im/contexts/settings/wallet/wallet_options/view.cljs]: Maybe we need to remove this `when` check as this will remove both the `Keypairs-and-accounts` and `saved-addresses` options. Additionally, we have the `when` check for `saved-addresses` in the basic settings list.

**Review by smohamedjavid — APPROVED** [review]
🚀 🚀 🚀

**Review by mohsen-ghafouri — APPROVED** [review]
Here we go 🚀

**Comment by pavloburykh** [comment]
@seanstrom thanks for the PR. Please take a look at the issues. 

### ISSUE 1 Save button is overlapped by keyboard on "Rename key pair" screen when Testnet banner is enabled (IOS)

**IOS ONLY** Reproducing on iPhone X, IOS 16.7.7

**Steps:** 

1. Enable Testnet
2. Open "Rename key pair" screen
3. Open keyboard
4. See if Save button is not overlapped by keyboard

**Actual result:** Save button is overlapped by keyboard

photo_2024-06-18 16 26 59

**Comment by seanstrom** [comment]
@pavloburykh thanks for finding that issue 🙌 

I've updated the rename key-pair screen to use the full-screen modal appearance, and I've made sure the button is visible when the keyboard is displayed too. Let me know if that resolves the issue please 🙏 

Here's a screen recording of the changes that I made:

https://github.com/status-im/status-mobile/assets/2845768/4848eff9-65d0-42dc-a43d-acaefecf8696

**Comment by pavloburykh** [comment]
Hey @seanstrom! thank you for fixing issue 1. 

I have a question related to importing [multiple missing key pairs](https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=232-32239&t=7XkHNDYUVGOFLIcr-0): does this feature supposed to work? There is no way to generate QR code for multiple keypairs as shown [here](https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=4791-50956&t=7XkHNDYUVGOFLIcr-0)

If this feature is not ready - then maybe we should hide "import" button? Is this feature tracked somewhere? 

Currently scanning QR code of separate keypair results in errors. And there is no way generate QR for multiple keypairs as I mentioned above. 

https://github.com/status-im/status-mobile/assets/97245802/49c24999-fd08-4ce8-b274-648dc8e47177

**Comment by pavloburykh** [comment]
### ISSUE 2 User is able to rename keypair using existing name

**Actual result:** 


https://github.com/status-im/status-mobile/assets/97245802/7ad6cbb1-8049-4813-be5c-f8f81a57b696

**Expected result:** validation error should be shown

https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=5562-16285&t=OXDrnyucz2wEFdpl-0

Descoped Wallet settings for Mobile – Figma 2024-06-19 15-10-15

**Comment by pavloburykh** [comment]
### ISSUE 3 Long keypair limitation does not match the design

@seanstrom for me it is not an issue, I like characters limit solution more than suggested by designers. But still, I am ponting this out as an issue because it differs from the design. 

**Actual result:** 


https://github.com/status-im/status-mobile/assets/97245802/b4cae0e0-8a30-4590-9930-e63b00894519

**Expected result:** according to Figma user is able to enter more than 20 characters which will result in validation error

https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=5562-16478&t=OXDrnyucz2wEFdpl-0

Descoped Wallet settings for Mobile – Figma 2024-06-19 15-15-01

**Comment by pavloburykh** [comment]
@seanstrom another question: In feature [description](https://github.com/status-im/status-mobile/issues/19936#issue-2285221964) I see the following:

```I want to be able to import a missing key pair by its private key```

At the same time I do not see such possibility neither in [designs](https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=232-33083&t=hioBwxRjpwpVBl41-0) nor in app. 

There are only 2 available options: import via seed phrase + import via QR code. 

Descoped Wallet settings for Mobile – Figma 2024-06-19 15-44-55


Do we expect a possibility to import via private key or there is a mistake in feature descriprion?

**Comment by pavloburykh** [comment]
> Currently scanning QR code of separate keypair results in errors

Overall, none of scan QR validation errors are implemented. So I am logging it as separate issue

### ISSUE 4 Scan QR validation errors do match the designs

**Steps:** 

1. When importing missing keypairs try scanning the following types of QRs: 

- QR code that belongs to the wrong key pair
- Already used keypair QR
- Invalid QR (non keypair QR)

2. Pay attention if validation errors match the designs. 

**Actual result:** non of the above cases is covered. Errors do match the designs. 

**QR code that belongs to the wrong key pair**
photo_2024-06-19 16 13 05

**Already used keypair QR**
photo_2024-06-19 16 10 22

**Invalid QR (non keypair QR)**
photo_2024-06-19 16 13 30


**Expected result:** 

https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=232-33031&t=hioBwxRjpwpVBl41-0

Descoped Wallet settings for Mobile – Figma 2024-06-19 16-09-29

**Comment by seanstrom** [comment]
> Hey @seanstrom! thank you for fixing issue 1.
> 
> I have a question related to importing [multiple missing key pairs](https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=232-32239&t=7XkHNDYUVGOFLIcr-0): does this feature supposed to work? There is no way to generate QR code for multiple keypairs as shown [here](https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=4791-50956&t=7XkHNDYUVGOFLIcr-0)
> 
> If this feature is not ready - then maybe we should hide "import" button? Is this feature tracked somewhere?
> 


Hey @pavloburykh thanks for finding this issue 🙌 

I double checked with the team and it seems that this import button should be feature-flagged since we don't currently have a way to export all the key-pairs in a single QR code. I'll add the feature-flag for this button soon 👍

**Comment by seanstrom** [comment]
> @seanstrom another question: In feature [description](https://github.com/status-im/status-mobile/issues/19936#issue-2285221964) I see the following:
> 
> `I want to be able to import a missing key pair by its private key`
> 
> At the same time I do not see such possibility neither in [designs](https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=232-33083&t=hioBwxRjpwpVBl41-0) nor in app.
> 
> There are only 2 available options: import via seed phrase + import via QR code.
> 
> Descoped Wallet settings for Mobile – Figma 2024-06-19 15-44-55
> Do we expect a possibility to import via private key or there is a mistake in feature descriprion?

Hey @pavloburykh 👋 

The designs for importing by a private-key should be here: https://www.figma.com/design/QWJV5WRbroXVDsdubEDjAT/Wallet-settings-for-Mobile?node-id=232-32719&t=xGA37vRLf3OvxIOl-4

And importing via a private-key will only be visible on key-pairs that have the "key" icon next to them. Here's an example of the two different action menus for a key-pair that's based on a private-key vs a key-pair that's based on a seed-phrase.

https://github.com/status-im/status-mobile/assets/2845768/4942e10d-d04d-4a68-9d53-b7d3d5b101c0

**Comment by pavloburykh** [comment]
### ISSUE 5 Seed phrase validation error issues

**IOS:** validation errors are not visible on IOS


https://github.com/status-im/status-mobile/assets/97245802/1a832709-2995-4bff-908c-9268ae218fcc


**Android:** validation errors are not visible when keyboard is closed


https://github.com/status-im/status-mobile/assets/97245802/9bb0658d-04f0-4549-bbdd-055f5884a74b

Also, slider does not return to it's initial position when using biometry (both IOS and Android)

photo_2024-06-19 16 29 16

**Comment by pavloburykh** [comment]
> And importing via a private-key will only be visible on key-pairs that have the "key" icon next to them. Here's an example of the two different action menus for a key-pair that's based on a private-key vs a key-pair that's based on a seed-phrase.

Thank you @seanstrom. But how can I generate such case with key pair that have the "key" icon next to it? We even do not have possibility to add new keypairs using private key. Wondering how you managed to generate such case.

**Comment by seanstrom** [comment]
> Thank you @seanstrom. But how can I generate such case with key pair that have the "key" icon next to it? We even do not have possibility to add new keypairs using private key. Wondering how you managed to generate such case.

Ah in this case I used a combination of status-desktop and a Metamask account:
* First I created an account on Metamask and copied the private key for that account.
* I then imported the private-key into status desktop with the status account that is synced to the mobile device.
* I then logged into the mobile device (you may need re-login to changes) and viewed the missing key-pairs.

The missing key-pairs should display the key-pair that's based on the private-key after those steps. Lmk if that works for you 🙏

**Comment by pavloburykh** [comment]
> > Thank you @seanstrom. But how can I generate such case with key pair that have the "key" icon next to it? We even do not have possibility to add new keypairs using private key. Wondering how you managed to generate such case.
> 
> Ah in this case I used a combination of status-desktop and a Metamask account:
> 
> * First I created an account on Metamask and copied the private key for that account.
> * I then imported the private-key into status desktop with the status account that is synced to the mobile device.
> * I then logged into the mobile device (you may need re-login to changes) and viewed the missing key-pairs.
> 
> The missing key-pairs should display the key-pair that's based on the private-key after those steps. Lmk if that works for you 🙏

Great, thank you @seanstrom!

**Comment by seanstrom** [comment]
@pavloburykh I've pushed up some fixes for Issues 2 & 3 related to validation when renaming a key-pair. Here's a screen recording of the changes 🎥 


https://github.com/status-im/status-mobile/assets/2845768/bc3532e6-3ebd-4e0e-bf41-f5e35d1142c8

**Comment by pavloburykh** [comment]
### ISSUE 6 "Could not decrypt key with given password" error when trying to perform transaction using key pair account imported via seedphrase/private key

**Steps:**

1. Sync Device A and Device B
2. Add new key pair via seedphrase/private key on Device A
3. Import new missing key pair on Device B using seedphrase/private key
4. On Device B try to perform transaction from newly imported key pair account
5. Observe the result: 

**Actual result:** cannot perform a transaction. After confirming a transaction "Could not decrypt key with given password" error appears. At the same time, in case of importing via Scanning QR code - transactions are performed successfully. 

[Status-debug-logs - 2024-06-20T124344.512.zip](https://github.com/user-attachments/files/15911931/Status-debug-logs.-.2024-06-20T124344.512.zip)

photo_2024-06-20 12 43 53

**Comment by pavloburykh** [comment]
### ISSUE 7 Import button opens camera despite there are no missing accounts which can be imported by syncing QR code

**Steps:** 

1. Create User A on Device A
2. Add new key pair by generating/adding via seedphrase/adding via private key (Desktop only)
3. Perform backup (manual back up can be performed on mobile only)
4. Restore User A on Device B
5. Go to Wallet settings and see missing keypairs
6. Tap import button
7. Observe the result

**Actual result:** import button opens camera. At the same time non of missing accounts can be imported by scanning QR code in this case. 


https://github.com/status-im/status-mobile/assets/97245802/ca685005-9ef5-4b89-b55b-07b7421f9bb2



**Expected result:** 

According to the [design](https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=4791-49761&t=vh4qlQIasg4cXjZY-0) import button should open entering seedphrase flow in case accounts are restored from Waku and there are no account which could be imported by scanning QR. 

At the same time, it is not clear from the designs how the flow with import button looks like in case of multiple accounts restored from Waku. It is also not covering the case when account should be imported via private key. I guess for such account import button should open Entering private key flow.  

image

**Comment by pavloburykh** [comment]
> I double checked with the team and it seems that this import button should be feature-flagged since we don't currently have a way to export all the key-pairs in a single QR code. I'll add the feature-flag for this button soon 👍

Thank you @seanstrom! As you see from [ISSUE 7](https://github.com/status-im/status-mobile/pull/20464#issuecomment-2180466566) "Import button" includes not only sync QR flows. 

Anyway, I agree it is right decision to feature flag it for now, as there are multiple issues related to this button.

**Comment by pavloburykh** [comment]
### ISSUE 8 "500 Internal server error when scanning key pair code

Please, follow exact steps for reproduction

**Preconditions:** User A has 1 or more additional key pairs that are backed up in Waku. 

**Steps:** 

1. Restore User A on Device A. In wallet settings you should see missing keypairs for those which are restored from Waku. 
2. Sync Device A with Device B
3. Import missing keypair (via seedphrase/private key) on Device B. On Device A this key pair will remain as missed. 
4. Try to import missing key pair on Device A by scanning QR code generated on Device B
5. Observe the result

**Actual result:** 

[Device_A_key_pair.zip](https://github.com/user-attachments/files/15916532/Device_A_key_pair.zip)
[Device_B_key_pair.zip](https://github.com/user-attachments/files/15916534/Device_B_key_pair.zip)


500 error on Device A

photo_2024-06-20 18 31 34

"Could not decrypt key with given password" on Device B

photo_2024-06-20 18 36 06

**Comment by pavloburykh** [comment]
### ISSUE 9 Cannot create derrived account using keypair which was imported after restoring from Waku

**Preconditions:** User A has 1 or more additional key pairs that are backed up in Waku. 

**Steps:**

1. Restore User A on Device A. In wallet settings you should see missing keypairs for those which are restored from Waku.
2. Import missing keypairs via seedphrase/private key
3. Try to create derrived account from imported key pair

**Actual results:** user gets stuck on confirmation page. From logs, I see the following error:

```cannot retrieve a valid key for a given account: could not decrypt key with given password```

[logs (35).zip](https://github.com/user-attachments/files/15925958/logs.35.zip)


https://github.com/status-im/status-mobile/assets/97245802/090408b0-2a23-4ed9-9005-402f7d7e1955

**Comment by seanstrom** [comment]
Hey @pavloburykh, thanks for finding these issues 🙌 

For issue 4:
* I've updated the code show the two of the designed toast error message
  * The "This is not the QR code you're looking for" error toast will appear if an import error occurs after scanning a valid QR code. To reproduce this error, we would scan a QR code for a different key-pair
  * The "This does not look like a key pair QR code" error toast will appear if a QR code is not related to a Status QR code. Though this error will not appear when scanning a sync QR from Status.
  * We're somewhat limited at the moment with being able to support these other QR code error messages, so I've attempted to implement 2 of 3 error messages where it seemed possible/appropriate.
    * cc @smohamedjavid 

For issue 5:
* I've updated the styles to ensure that the validation error messages are visible on iOS.
* I've updated the styles and layout for the recovery-phrase form to also show the validation error messages when the keyboard is dismissed.
* And I've updated the code to ensure that the slider for biometric auth is reset to the starting position.

For issue 7:
* I've added a feature-flag to remove the "import" button from the missing key pairs UI.

For issues 6, 8 and 9:
* I've updated the import logic that should resolve these issues with `500` errors and password decryption errors. Can you re-test these with the latest updates?

**Comment by pavloburykh** [comment]
@seanstrom thank you for the fixes. I would like to include https://github.com/status-im/status-mobile/pull/20469 in this PR and continue testing then. Once @smohamedjavid will merge it, let's rebase current PR. Thank you!

**Comment by seanstrom** [comment]
@pavloburykh I've updated this PR with the latest changes from develop. I accidentally merged the branch, but I’ll clean that up once I’m back at my desk. But the PR should be ready to test 👍

edit:
actually never mind haha, I’ll update this PR in a bit, my mobile GitHub isn’t working quite right

**Comment by pavloburykh** [comment]
@seanstrom thank you! Also, I see integration tests are constantly failing https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20464/15/consoleText

**Comment by smohamedjavid** [comment]
> @seanstrom thank you! Also, I see integration tests are constantly failing https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20464/15/consoleText

@pavloburykh - Fixed now 👍

**Comment by qoqobolo** [comment]
Hey @seanstrom, thanks for the fixes!
Found some new issues but please lmk if some of them are not blockers for this PR and can/should be fixed separately 
cc @pavloburykh 


### ISSUE 10: Error when opening `Key pairs and accounts` right after recovering the profile if there is a missing key pair

After re-opening the app it works fine. 

**Steps:**
1. Create a profile
2. Create a new key pair
3. Perform backup
4. Recover this user on another device using seed phrase
5. Without reopening the app go to `Profile` > `Wallet` > `Key pairs and accounts`

Screenshot 2024-06-24 at 12 58 06

### ISSUE 11: Wrong error text when scanning a QR from a wrong key pair when attempting to import a missing key pair

**Steps:**
1. Create a profile on Device1
2. Sync Device2 with Device2
3. Create a new key pair on Device1
4. On Device1: open `Profile` > `Wallet` > `Key pairs and accounts` > press three dots on some key pair, other than needed one (e.g. on the default key pair) > `Show encrypted QR of key pairs on device`
5. On Device2: press three dots on the missing key pair > `Import by scanning encrypted QR` > Scan the wrong QR

**Expected error:** `This is not the QR you are looking for` (pls correct me if I'm wrong here)
**Actual error:** 

Screenshot 2024-06-24 at 12 59 25

### ISSUE 12: _Presumably_ wrong error text when scanning an already used QR

This might be an edge case, but I honestly don't know how else to reproduce the "QR code has already been used" error, so I tried this scnario

**Steps:**
1. Create a profile on Device1
2. Sync Device2 and Device3
3. Create a new key pair on Device1
4. Scan the required QR code for the missing key pair on Device2
5. Scan **_the same_** QR code for the missing key pair on Device3

**Expected error:** `QR code has already been used`
**Actual error:**

Screenshot 2024-06-24 at 13 25 35


### ISSUE 13: [Android] Extra space above the keyboard on the seed phrase screen

The issue is **not** reproducible on iOS;
on Android I checked it on Pixel 7a

Screenshot 2024-06-24 at 13 20 04

**Comment by qoqobolo** [comment]
@seanstrom UPD (in case you've already seen my comment above🙂 ): just removed ISSUE 10 (was the first in the list) since it's already logged https://github.com/status-im/status-mobile/issues/20416 and updated issue nimbers

**Comment by seanstrom** [comment]
@qoqobolo thanks for finding these issues 🙌 

Here's some notes on the issues:
* I'll start looking into issues 10 and 13, I think those should be easy to fix 👍 
* For issues 11 and 12, I'm not sure if I can implement these correct errors messages. But I will try finding the places to emit the correct toast messages. Though I'm not sure if there's a way to check if we've already used a QR code 🤔 
  * cc @smohamedjavid

**Comment by qoqobolo** [comment]
> For issues 11 and 12, I'm not sure if I can implement these correct errors messages. But I will try finding the places to emit the correct toast messages. Though I'm not sure if there's a way to check if we've already used a QR code 🤔

Sure, please let me know if we need to log this separately or if you have more details about how exactly we can reproduce those errors. 
Thank you @seanstrom!

**Comment by smohamedjavid** [comment]
@qoqobolo - Can you help me create follow-up issues for 11 and 12 (Initial Issue 4 in this PR)? 🙏 We will look into those QR-related issues (unhappy flows) separately. For Issue 12, we are not sure about knowing whether the QR is already used. I will check that part as the whole import missing key pair using QR is using the local pairing features to transfer the keystore files of key pairs. 👍

**Comment by seanstrom** [comment]
@qoqobolo Oki doki 🧑‍🌾 
I've pushed up some changes for fixing issues 10 and 13, please let me know if those are resolved 🙏

**Comment by qoqobolo** [comment]
> @qoqobolo - Can you help me create follow-up issues for 11 and 12 (Initial Issue 4 in this PR)? 🙏 We will look into those QR-related issues (unhappy flows) separately. For Issue 12, we are not sure about knowing whether the QR is already used. I will check that part as the whole import missing key pair using QR is using the local pairing features to transfer the keystore files of key pairs. 👍

@smohamedjavid sure thing, will do 👍

---
@seanstrom thanks for the fixes, LGTM now! 
I have one more question:

### ISSUE 14/Q: What validation error for a key pair name should we show when it's too short?

Currently, we show `Your key pair name is too short`

Screenshot 2024-06-24 at 17 01 29

But the thing is that there is no such error in the [desings](https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=5562-16285&t=MaZbjQ2BbdXCyPGk-4):

Screenshot 2024-06-24 at 17 07 10

Screenshot 2024-06-24 at 17 07 21

Also, we do not prompt the user through the UI for a minimum name length (like we do with a name that's too long by making the character count red).
Perhaps we should use a validation error that we use when creating a key pair - `Key pair name must be at least 5 characters`? Unless, of course, there is a previous discussion behind this decision.

Screenshot 2024-06-24 at 17 01 42

**Comment by pavloburykh** [comment]
@seanstrom thank you for the fixes. 

> I've pushed up some changes for fixing issues 10 and 13, please let me know if those are resolved 

ISSUE 13 is resolved for me. 

ISSUE 10 is still reproducible. I propose to keep it track and fix separately in order not to block this PR. WDYT? 

**UPDATE:** @seanstrom sorry for confusion, regrading ISSUE 10 I thought you have claimed to fix the issue which has been removed by Diana. Regarding ```ISSUE 10: Error when opening Key pairs and accounts right after recovering the profile if there is a missing key pair``` - it is fixed. So we are good))

**Comment by seanstrom** [comment]
> Perhaps we should use a validation error that we use when creating a key pair - Key pair name must be at least 5 characters? Unless, of course, there is a previous discussion behind this decision.

@qoqobolo Thanks for spotting this 🙌 
I think it makes sense to update validation error message to be consistent, I'll make that change 👍

**Comment by seanstrom** [comment]
Oki doki I've updated validation error message 🌳 
Let me know if there any other issues 🙏

**Comment by qoqobolo** [comment]
Everything looks good to me now, PR is ready to be merged 🎉🎉
@seanstrom @pavloburykh thank you guys and everyone involved for the amazing job you've done!

**Comment by qoqobolo** [comment]
Does this PR need a design review, btw?🤔
Please ping Francesca if so 🙏

**Comment by seanstrom** [comment]
@qoqobolo yup yup this needs a design review 🎨 

@Francesca-G can you please review this PR 🙏

**Comment by Francesca-G** [comment]
> @Francesca-G can you please review this PR 🙏

Looks good so far, just to make sure, is this feature yet to be added? 

Screenshot 2024-06-25 alle 11 07 31

>
This should be showing up when pressing the _options_ button in the **Key pairs and accounts** screen

**Comment by seanstrom** [comment]
> Looks good so far, just to make sure, is this feature yet to be added?
> 
> This should be showing up when pressing the _options_ button in the **Key pairs and accounts** screen

@Francesca-G Thanks for noticing this 🙏, I double checked on my end, and I think that feature should be implemented, but it might require the account to have at least 2 paired / synced devices. Does that device have another device paired / synced with the account?

**Comment by qoqobolo** [comment]
> > @Francesca-G can you please review this PR 🙏
> 
> Looks good so far, just to make sure, is this feature yet to be added?
> 
> Screenshot 2024-06-25 alle 11 07 31
> > 
> 
> This should be showing up when pressing the _options_ button in the **Key pairs and accounts** screen

@Francesca-G hmm as I can see on the screenshot this is a default key pair, and you're right, according to the [design](https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=49-133839&t=IOvsRbbJUFnYgf7W-4) we should show QR even in this case.

But I'm wondering if this is actually the case. I mean is such a scenario really possible when the default key pair (`on device`) will be a `missing key pair` and we will have to show a QR code for it?

**Comment by Francesca-G** [comment]
> @Francesca-G Thanks for noticing this 🙏, I double checked on my end, and I think that feature should be implemented, but it might require the account to have at least 2 paired / synced devices. Does that device have another device paired / synced with the account?

No, it doesn't

Screenshot 2024-06-25 alle 11 21 23

**Comment by qoqobolo** [comment]
Maybe we should also hide this three dots button for default key pair, wdyt? @seanstrom @Francesca-G 
Screenshot 2024-06-25 at 11 22 01

**Comment by qoqobolo** [comment]
@Francesca-G just in case sharing here the steps you need to take to see the `Show encrypted QR of key pairs on device` so that you can test it:

1. Create a profile on Device1
2. Sync Device2 with Device1
3. Create a new key pair on Device1
4. Open `Key pairs and accounts` on Device1 and open this three dots menu for the newly created key pair

**Comment by seanstrom** [comment]
> Maybe we should also hide this three dots button for default key pair, wdyt? @seanstrom @Francesca-G Screenshot 2024-06-25 at 11 22 01

Yeah good question! I'm not sure what the purpose is of the default status account, but if we create one for each device, then maybe we want to share the QR code to other devices. cc @smohamedjavid do you have more context?

**Comment by seanstrom** [comment]
I think atm when we don't have a paired device we cannot see the that "Show QR" action on any of the key-pairs. I'm not familiar with the exact details as to why, but perhaps @smohamedjavid would have more context if we have more questions.

**Comment by seanstrom** [comment]
@Francesca-G are you able to see the "Show QR" action after following those steps from @qoqobolo?

**Comment by mohsen-ghafouri** [comment]
Yes as you already mentioned, this option is only available if you have a paired devices, https://github.com/status-im/status-mobile/pull/20073

also another option for missing keypair, "Import by scanning encrypted QR" is only available when you have paired devices

**Review by Francesca-G — APPROVED** [review]
yes, @seanstrom all good now

as suggested by @qoqobolo we could remove the options button in case there's only one account

**Comment by seanstrom** [comment]
> yes, @seanstrom all good now

Awesome! 🙌 

> as suggested by @qoqobolo we could remove the options button in case there's only one account

Ah I see, I misunderstood the suggestion, I can do that now 👍

**Comment by smohamedjavid** [comment]
Thanks, @qoqobolo for testing the PR and @Francesca-G for the design review 🙏 

> Looks good so far, just to make sure, is this feature yet to be added?
> 
> Screenshot 2024-06-25 alle 11 07 31
> > 
> 
> This should be showing up when pressing the _options_ button in the **Key pairs and accounts** screen

This shouldn't be the case, we don't actually need to export the `profile` key pair via QR as your other paired device is already signed in with your profile. There is no way the profile key pair would be missing. Additionally, the desktop app shows only an option to move the key pair to the keycard and NOT to export it.

We might need to remove the three dots for the profile key pair as @qoqobolo suggested, until we implement the keycard features for moving it.

cc: @xAlisher

**Comment by xAlisher** [comment]
Screenshot is not available.

**Comment by smohamedjavid** [comment]
> Screenshot is not available.

@xAlisher - The options for profile keypair
https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=49-133839&t=pFdA25FTTS4KPgk0-1

**Pull request merged**
